File: 211_Add_and_Search_Word_Data_structure_design.cpp
#include<bits/stdc++.h>

using namespace std;

class WordDictionary {
public:
    /** Initialize your data structure here. */
    struct TrieNode{
        char letter;
        bool WordEnd = false;
        TrieNode *children[26];
    };
    
    TrieNode *root = new TrieNode();
    
    /** Adds a word into the data structure. */
    void addWord(string word) {
        TrieNode *p = root;
        for (const char a : word){
            int idx = a - 'a';
            if (!p->children[idx]){
                p->children[idx] = new TrieNode();
                p->children[idx]->letter = a;
            }
            p = p->children[idx];
        }
        p->WordEnd = true;
    }
    
    /** Returns if the word is in the data structure. A word could contain the dot character '.' to represent any one letter. */
    bool search(string word) {
        return SearchWord(word, root, 0);
    }
    
    bool SearchWord(string word, TrieNode *head, int s){
        if (s >= word.length()) return head->WordEnd;
        TrieNode *p = head;
        for (int i = s; i < word.length(); ++i){
            if (word[i] == '.') {
                bool res = false;
                for (auto child : p->children) {
                    if (child) res = res || SearchWord(word, child, i + 1);
                }
                return res;
            }
            int idx = word[i] - 'a';
            if (!p->children[idx]) return false;
            p = p->children[idx];
        }
        return p->WordEnd;
    }
};


int main(){
	WordDictionary* worddic = new WordDictionary();
	worddic->addWord("bad");
	worddic->addWord("dad");
	worddic->addWord("mad");
	cout << worddic->search("pad") << endl;
	cout << worddic->search("bad") << endl;
	cout << worddic->search(".ad") << endl;
	cout << worddic->search("b..") << endl;
    return 0;
}

File: 211_Add_and_Search_Word_Data_structure_design.py
class TrieNode():
    def __init__(self):
        self.children = [None] * 26
        self.endWord = False

class WordDictionary(object):
    """ O(word) : 20% """
    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.root = TrieNode()

    def addWord(self, word):
        """
        Adds a word into the data structure.
        :type word: str
        :rtype: None
        """
        p = self.root
        for c in word:
            idx = ord(c) - ord('a')
            if not p.children[idx]:
                p.children[idx] = TrieNode()
            p = p.children[idx]
        p.endWord = True

    def search(self, word):
        """
        Returns if the word is in the data structure. A word could contain the dot character '.' to represent any one letter.
        :type word: str
        :rtype: bool
        """
        p = self.root
        return self.dfs_search(word, p)

    def dfs_search(self, word, root):
        res = False
        if not word:
            if root.endWord: res = True
            return res

        if word[0] != '.':
            idx = ord(word[0]) - ord('a')
            if root.children[idx] is None:
                return res
            res = self.dfs_search(word[1:], root.children[idx])
        else:
            for child in root.children:
                if child is None:
                    continue
                res = self.dfs_search(word[1:], child)
                if res: break
        return res



# Your WordDictionary object will be instantiated and called as such:
obj = WordDictionary()
obj.addWord("bbb")
print obj.search("b..")